File: projects/engineering.purdue.edu/programmingwithobjects/pwocode/chap10/TryCatch.cc
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:  Avinash Kak
//
// Chapter:     Chapter 10  Handling Exceptions
//
// Section:     Section 10.2  Exception Handling In C++
//


//TryCatch.cc

#include <iostream>
#include <cstdlib>
using namespace std;

void f( int );

class Err {};                                                     //(A)

int main()
{
    try {                                                         //(B)
        f(0);                                                     //(C)
    } catch( Err ) {                 
        cout << "caught Err" << endl; 
        exit(0);
    }
 
    return 0;
}

void f(int j) {
    cout << "function f invoked with j = " << j << endl;
    if (j == 3) throw Err();                                      //(D)
    f( ++j );
}

File: projects/engineering.purdue.edu/programmingwithobjects/pwocode/chap10/TryCatch.java
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:   Avinash Kak
//
// Chapter:     Chapter 10 Handling Exceptions
//
// Section:     Section 10.4  Differences Between C++ And Java For Exception Handling
//



//TryCatch.java

import java.io.*;

class Err extends Exception { }

class Test {
    public static void main( String[] args )
    {
        try {
            f(0);
        } catch( Err e ) {
            System.out.println( "Exception caught in main" );     
        }
    }
  
    static void f(int j) throws Err {                             
        System.out.println( "function f invoked with j = " + j );
        if (j == 3) throw new Err();                              
        f( ++j );
    }
}